### DISCUSSION TITLE
An option to resolve all conflicts at runtime

### DISCUSSION THREAD

**Discussion by CyberShadow** [tree-sitter/tree-sitter#4702] [Ideas]
The work-in-progress tree-sitter grammar for the D programming language that I am working on currently has [198 conflicts](https://github.com/cybershadow/tree-sitter-d/blob/master/grammar.js#L23-L270).

Perhaps some of these should not be listed there, and instead should be resolved using associativity or precedence rules or such. Indeed, [ideally I or someone else would go through all of them and make an informed decision for how best to handle the LR(1) conflict](https://github.com/CyberShadow/tree-sitter-d/issues/1). However, doing so for even one is difficult and time consuming, as it requires studying the language specification beyond the BNF-like grammar, possibly even looking at the implementation, so for now I've been shoving all conflicts into the `conflicts` list. It's been working OK so far.

More importantly, the requirement to enumerate all LR(1) conflicts poses a hard blocker for being able to automatically convert the grammar from the language's [official specification](https://dlang.org/spec/intro.html) to tree-sitter. I've been hoping to use tree-sitter as part of an automatic CI system, in which changes to the grammar would be automatically tested against the compiler's test suite, and vice versa. Because new tree-sitter conflicts can arise due to changes in the language specification, the process cannot be fully automated (without parsing tree-sitter's error message, which are meant for humans and not machines).

So, my options are:

1. Wastefully reimplement the algorithm that tree-sitter uses to construct the parsing tables and conflict resolution, in order to generate a full and precise `conflicts` array which would match `tree-sitter`'s expectations.
2. Parse `tree-sitter`'s error messages and re-run it 198 times in order to thus build a `conflicts` array as part of CI.
3. Propose a change in `tree-sitter` to allow writing e.g. `conflicts: '*'`, which would mean to automatically use GLR for all LR(1) conflicts, as if the `conflicts` array contained all possible rule combinations.

Or, does a better solution exist?

Implementing option 3 in tree-sitter should be trivial, so I'd be happy to submit a pull request.

**Comment by maxbrunsfeld** [comment]
Generally, language specifications are very different from the kinds of formal grammars that can be used to generate efficient parsers. In my experience, language specs are for humans to read. They aren't designed to be unambiguous using a finite number of lookahead tokens.

My recommendation would be to not try to generate a parser from the spec. If you want your parser to be useful and efficient, you should hand-code the tree-sitter grammar.

**Comment by CyberShadow** [comment]
Well, the grammar is actually already almost fully working: https://cy.md/d/tree-sitter-d/

> My recommendation would be to not try to generate a parser from the spec. If you want your parser to be useful and efficient, you should hand-code the tree-sitter grammar.

This is not realistic. D is a very large language, which is still evolving; a hand-written grammar for it would be outdated by the time it's finished. That's not what my problem is about, though.

I'm trying to use tree-sitter for more than just parsing text in an editor: I'd like to use it as a second partial implementation of the language based on the spec itself, to cross-verify the primary implementation (using its test suite). It's going mostly well, with the problem described here being the biggest currently known blocker.

**Comment by CyberShadow** [comment]
@maxbrunsfeld If you'd like to learn more about my project for more background information for this issue, check the README here: https://github.com/CyberShadow/tree-sitter-d :)

**Comment by CyberShadow** [comment]
(I would also be OK with a "hard no" for this feature request; I would probably just use a tree-sitter fork which generates instead of requiring the `conflicts` array as part of CI, but official releases of the grammar would have the `conflicts` array generated by the fork.)

**Comment by maxbrunsfeld** [comment]
I see, so you aren't necessarily trying to create a generally-useful D parser, and this is more of a niche project for testing D's compiler and its spec? I don't want to make your job harder, but I'm just trying to clarify what the ideal solution would be, as it concerns Tree-sitter.

Even for a special-purpose parser like you're describing, I don't think that resolving all conflicts at runtime is going to serve your purposes well. In many cases, multiple conflicting interpretations will be grammatically valid, and Tree-sitter is going to pick one arbitrary in the final syntax tree. Of course, you can control which one it picks by adding *dynamic precedence*, but then you're back to manually editing your grammar.

Does the spec provide information about the precedences of different operators? I'd think you'd at least need to supply that information to Tree-sitter to get anything predictable out of it. And once you add that, you're already having to incorporate some non-BNF information into your grammar.

**Comment by CyberShadow** [comment]
> I see, so you aren't necessarily trying to create a generally-useful D parser, and this is more of a niche project for testing D's compiler and its spec?

Well, I would like to attempt to do both, insofar as this direction carries me, and it's working pretty well so far.

> Even for a special-purpose parser like you're describing, I don't think that resolving all conflicts at runtime is going to serve your purposes well. In many cases, multiple conflicting interpretations will be grammatically valid, and Tree-sitter is going to pick one arbitrary in the final syntax tree. Of course, you can control which one it picks by adding _dynamic precedence_, but then you're back to manually editing your grammar.

I understand. Still, for all tests so far, I have not encountered any problems.

In fact, so far I hadn't noticed any situation that could be improved by specifying precedence or associativity in the tree-sitter grammar. I think the large number of conflicts is simply due to the language requiring lots and lots of lookahead, in many situations.

> Does the spec provide information about the precedences of different operators?

You mean for the infix binary operators? Yes, those are done with e.g. `OrOrExpression` expanding to `AndAndExpression` etc. I remove the intermediate rules from the tree-sitter tree by splitting them into an anonymous part and a named part.

**Comment by CyberShadow** [comment]
> In my experience, language specs are for humans to read.

I should probably clarify that I've been also making improvements from the other end - this project has already led to [many fixes](https://github.com/CyberShadow/tree-sitter-d/issues/4) to the specification, a few for the compiler test suite, and even two to change the implementation to conform to the spec.

And, FWIW, at the last community meeting the language creator encouraged the effort of making the official specification machine-readable as well :)

**Comment by stephe-ada-guru** [comment]
Vladimir Panteleev ***@***.***> writes:

> 2. Parse `tree-sitter`'s error messages and re-run it 198 times in
> order to thus build a `conflicts` array as part of CI.

For whatever it's worth, the conflict messages output by the wisitoken
grammar generator are in the correct format to be pasted into the
grammar source file as conflict declarations, and it lists all conflicts
in one pass. It might be best if tree-sitter did the same.

-- 
-- Stephe

**Comment by TamaMcGlinn** [comment]
Okay, since the best option for this is still to parse tree-sitter generate's error messages, [here you go](https://github.com/TamaMcGlinn/tree-sitter-ada/blob/master/conflict_generator):

```
#!/usr/bin/env bash
# This repeatedly runs `tree-sitter generate` to get each conflict,
# adds it to the list of conflicts in grammar.js and re-runs until there are no conflicts
# as discussed here, this feature should be added to tree-sitter itself:
# https://github.com/tree-sitter/tree-sitter/issues/1365

# To use, ensure that grammar.js has a list of conflicts like this:
# conflicts: $ => [
#   [optional other items already ending in comma],
#   // insertion marker
# ],
# the insertion marker specifies the indentation that will be used

add_conflict(){
  echo "Inserting $1"
  sed -i "s&\( *\)\(// insertion marker\)&\1$1\n\1\2&" grammar.js 
}
run_ts(){
  tree-sitter generate &> last_ts_output
}
run_ts
LINE=$(grep "Add a conflict" last_ts_output)
while [ $? == 0 ]
do
  CONFLICT=$(echo $LINE | sed 's/.*: `/[$./' | sed 's/`$/],/' | sed 's/`,/,/g' | sed 's/`/$./g')
  echo "Found conflict: $CONFLICT"
  add_conflict "$CONFLICT"
  run_ts
  LINE=$(grep "Add a conflict" last_ts_output)
done
```
